### PR TITLE
Move code from eden related to adding pre-commit hooks.

### PR DESCRIPTION
**Pull request by fabioz**
In this pull request I've added code to add pre-commit hooks (moved from eden), improved docs (so that they no longer reference so much internal things from esss) and bumped up the version.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -150,8 +135,8 @@ Create a conda environent (using Python 3 here) and install it in development mo
 
     $ conda create -n esss-fix-format-py3 python=3
     $ source activate esss-fix-format-py3
-    $ pip install -e .
     $ pip install . -r requirements_dev.txt
+    $ pip setup.py develop
```
Comment (nicoddemus) [inline README.rst]: Typo: `python setup.py develop`.

But actually `pip install -e .` is recommended over `python setup.py develop`.

I suggest to install this with `requirements_dev.txt` then:

```
pip install -e . -r requirements_dev.txt
```

File: esss_fix_format/cli.py [hunk 2]
```diff
@@ -30,24 +30,57 @@ def should_format(filename):
...
+    if git_hooks:
+        from esss_fix_format.hook_utils import install_pre_commit_hook
+        install_pre_commit_hook()  # uses the current directory by default.
+        return
+
     formatter = []
 
     def format_code(code_to_format):
         if not formatter:
+            # Note: before actually starting a new process, make sure the previous have been
```
Comment (nicoddemus) [inline esss_fix_format/cli.py]: Please move this code to a separate function, something like `_wait_current_processes(timeout=0.05)` or so. You can then change the comment to the function's docstring.

File: esss_fix_format/hook_utils.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        begin_in = os.curdir
+
+    base_directory = os.path.abspath(begin_in)
+    while True:
+        directory = base_directory
+        if os.path.exists(os.path.join(directory, filename)):
+            return directory
+
+        parent_base_directory, current_dir_name = os.path.split(base_directory)
+        if len(current_dir_name) == 0:
```
Comment (nicoddemus) [inline esss_fix_format/hook_utils.py]: `if not current_dir_name:` is considered more Pythonic

File: esss_fix_format/hook_utils.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            return None
+        if len(parent_base_directory) == 0:
```
Comment (nicoddemus) [inline esss_fix_format/hook_utils.py]: `if not parent_base_directory:` is considered more Pythonic

File: esss_fix_format/hooks.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            return meta(name, bases, d)
+
+        @classmethod
+        def __prepare__(cls, name, this_bases):
+            return meta.__prepare__(name, bases)
+
+    return type.__new__(metaclass, str('temporary_class'), (), {})
+
+
+class GitHook(with_metaclass(abc.ABCMeta, object)):
```
Comment (prusse-martin) [inline esss_fix_format/hooks.py]: Why do we need a metaclass here?

File: esss_fix_format/hooks.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                echo "fix-format check failed (status=$returncode)! To fix, execute:"
+                echo "  ff -c"
+                exit 1
+            fi
+        fi
+        """
+        return textwrap.dedent(script)
+
+
+class FdescribeGitHook(GitHook):
```
Comment (prusse-martin) [inline esss_fix_format/hooks.py]: This hook is really not used?

File: esss_fix_format/hooks.py (hunk 5 above)
Comment (fabioz) [reply esss_fix_format/hooks.py]: I have no idea -- just moved the code from eden ;)

Comment (prusse-martin) [reply esss_fix_format/hooks.py]: hehehe.
OK.

I guess the original author was using if for [abc.abstractmethod](https://docs.python.org/2/library/abc.html#abc.abstractmethod).

File: esss_fix_format/hooks.py (hunk 6 above)
Comment (fabioz) [reply esss_fix_format/hooks.py]: It seems it's used in rfdap... I'll remove the version from `esss_fix_format` (it doesn't seem to make much sense here) -- we can keep the version available in eden.

File: README.rst (hunk 1 above)
Comment (fabioz) [reply README.rst]: Changed.

**Pull request merged**